src/payment/strategies/adyenv2/adyenv2-payment-strategy.ts

```diff
@@ -0,0 +1,289 @@
+                                    .then((payment: Payment) =>
+                                        this._store.dispatch(this._paymentActionCreator.submitPayment(payment)));
+                            }
+
+                            return Promise.reject(error);
+                        });
+                }
+
+                if (error.body.three_ds_result.result_code === ResultCode.ChallengeShopper) {
+
+                    return this._handle3DS2Challenge(error.body.three_ds_result, payment.methodId)
+                        .then((payment: Payment) =>
+                            this._store.dispatch(this._paymentActionCreator.submitPayment(payment)));
+                }
+
+                return new Promise(() => {
+                    this._formPoster.postForm(error.body.three_ds_result.acs_url, {
+                        PaReq: error.body.three_ds_result.payer_auth_request,
+                        TermUrl: error.body.three_ds_result.callback_url,
+                        MD: error.body.three_ds_result.merchant_data,
+                    });
+                });
+            });
+    }
+
+    finalize(): Promise<InternalCheckoutSelectors> {
+        return Promise.reject(new OrderFinalizationNotRequiredError());
+    }
+
+    deinitialize(): Promise<InternalCheckoutSelectors> {
+        if (this._adyenOldComponent) {
+            this._adyenOldComponent.unmount();
+            this._adyenOldComponent = undefined;
+        } else if (this._adyenComponent) {
+            this._adyenComponent.unmount();
+            this._adyenComponent = undefined;
+        }
+
+        return Promise.resolve(this._store.getState());
+    }
+
+    private _getAdyenV2PaymentInitializeOptions(): AdyenV2PaymentInitializeOptions {
+        if (!this._adyenv2) {
+            throw new InvalidArgumentError(' "options.adyenv2" argument was not provided during initialization.');
+        }
+
+        return this._adyenv2;
+    }
+
+    private _getLocale(): string {
+        const state = this._store.getState();
+        const storeConfig = state.config.getStoreConfig();
+
+        if (!storeConfig) {
+            return 'en_US';
+        }
+
+        return storeConfig.storeProfile.storeLanguage;
+    }
+
+    private _getStateContainer(): string {
+        if (!this._stateContainer) {
+            return '{}';
+        }
+
+        return this._stateContainer;
+    }
+
+    private _getThreeDS2ChallengeWidgetSize(): string {
+        const { threeDS2ChallengeWidgetSize } = this._getAdyenV2PaymentInitializeOptions();
+
+        if (!threeDS2ChallengeWidgetSize) {
+            return '01';
+        }
+
+        return threeDS2ChallengeWidgetSize;
+    }
+
+    private _handle3DS2Challenge(resultObject: ThreeDS2Result, paymentMethodId: string): Promise<Payment> {
+
+        return new Promise((resolve, reject) => {
+            if (!this._adyenCheckout) {
+                throw new MissingDataError(MissingDataErrorType.MissingPaymentMethod);
```

Comment:
This might not be the most suitable error to throw. If you look at the error message, it says "Unable to proceed because payment method data is unavailable or not properly configured." This is not entirely true. I think a more suitable error to throw is `NotInitializedError` because one of the main reasons why `_adyenCheckout` is undefined is because the Adyen JS client is not initialized.